docs/testing/02-cli-tests.mdx

```diff
@@ -0,0 +1,123 @@
+}
+```
+
+### Defining Individual Test Cases
+
+Within the `test_cases` array, you can define individual test cases. Each test case can
+define the following fields:
+
+- `name`: The name of the test case, which will be used to identify the test case if it fails.
+- `database`: The name of the database schema that you are targeting, as defined at the top of the schema.
+- `target`: The name of the action/procedure in the schema that you were testing.
+- `args`: An array of arguments that will be passed to the procedure. It must match the expected parameters of the procedure. It can be empty if the procedure does not take any arguments.
+- `returns`: A 2-dimensional array of expected return values. This matches the expected return table of the action/procedure. This can be empty if the procedure does not return anything, or if an error is expected.
+- `error`: If the test case is expected to return an error, you can define the expected error message here. The tool will check that the returned error contains this string as a substring. If empty, the test case is expected to succeed.
+- `caller`: The wallet address (or any other string) that will be assigned to `@caller`. If empty, the caller will be set to the same wallet address as the database deployer.
+- `height`: The height of the block that the test case will be run at. If empty, the height will be set to 0.
+
+Below, we add two test cases to our schema. The first test case tests the `create_user` procedure, and the second tests the `get_users` procedure.
+**All data will be wiped after each test case is run,** except for the data that was seeded in the `seed_statements` field.
+
+```json title="tests.json"
+{
+    "name": "my users.kf tests",
+    "schema_files": ["users.kf"],
+    "seed_statements": {
+        "users_db": ["INSERT INTO users (id, name, address) VALUES ('42f856df-b212-4bdc-a396-f8fb6eae6901'::uuid, 'satoshi', '0xAddress')"]
+    },
+    "test_cases": [
+        {
+            "name": "create_user test",
+            "database": "users_db",
+            "target": "create_user",
+            "args": ["alice"],
+            "caller": "0xOtherAddress"
+        },
+        {
+            "name": "get_users test",
+            "database": "users_db",
+            "target": "get_users",
+            "args": [],
+            "returns": [["satoshi", "0xAddress"]]`
+        }
+    ]
+}
+```
+
+## Running Tests
+
+To run the test, use the `kwil-cli utils test` command. The test file path can be passed using
+the `--file` flag. Multiple test files can be spcified by simply adding more `--file` flags.
+The test also needs a Postgres instance to run against. You can use the `--test-container` flag
+to have the test automatically start a Postgres instance in a Docker container. Alternatively, you
+can connect to your own Postgres image using the `--host`, `--port`, `--user`, `--password`, and `--database` flags.
+
+
+```shell
+kwil-cli utils test --file tests.json --test-container
```

Comment:
Note: If docker daemon is not running, it does not elevate the docker error (usually gives a helpful hint like: `Is Docker Daemon running?`). Is it possible to elevate the docker errors when running the tests?

Reply:
Yeah. This requires a change to the repo, but I can make it.

Reply:
It appears this is already added:

```
$ kwil-cli utils test --file testing/proxy/impl_2_test.json --test-container
1.7210642454402392e+09  info    testing/testing.go:536  using schema "impl_1" (DBID: "x9f6687895d32a12f4c65e8ad847fb3a1e8eb3bc38524e5b8a0475a1b")
1.721064245443011e+09   info    testing/testing.go:536  using schema "impl_2" (DBID: "xa0ac96f1bc96d0e6e1ed233b1e8de62e218b6002c152c71877f38187")

Tests failed:
error running test container: failed to run ["docker" "run" "-d" "-p" "52853:5432" "--name" "kwil-testing-postgres" "-e" "POSTGRES_HOST_AUTH_METHOD=trust" "kwildb/postgres:latest"]:
docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
See 'docker run --help'.
```